warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownNegationVisitor.java

```diff
@@ -0,0 +1,256 @@
+import org.apache.commons.jexl2.parser.ASTOrNode;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.JexlNodes;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Pushdown negations until they are either wrapping leaf nodes or bounded ranges. The purpose of this class is to remove edge cases from evaluating
+ * executability with respect to negations.
+ *
+ * Visitor requires a flattened tree prior to execution or DeMorgans law will not be applied correctly. Negations will never be pushed down into EQ/ER nodes.
+ * Negations will be pushed down to flip NEQ and NR nodes
+ */
+public class PushdownNegationVisitor extends BaseVisitor {
+    
+    /**
+     * Create a copy of the tree and flatten it, then pushdown all negations
+     * 
+     * @param tree
+     *            the original tree
+     * @return the modified tree with all negations pushed down
+     */
+    public static JexlNode pushdownNegations(JexlNode tree) {
+        // flatten the input
+        JexlNode flattened = TreeFlatteningRebuildingVisitor.flatten(tree);
+        
+        flattened.jjtAccept(new PushdownNegationVisitor(), new NegationState(false));
+        
+        // flatten the result
+        return TreeFlatteningRebuildingVisitor.flatten(flattened);
+    }
+    
+    @Override
+    public Object visit(ASTNotNode node, Object data) {
+        // if there is already a negation passed down it should negate this node and continue down the tree to check for other negations
+        if (data != null && ((NegationState) data).isNegated()) {
+            JexlNode nextNode = dropNot(node, (NegationState) data);
+            
+            // pass down with a new negation state since nothing that comes up from this point should impact what is already being returned (a processed
+            // negation)
+            nextNode.jjtAccept(this, new NegationState(false));
+            
+            return data;
+        } else {
+            // arrived to this node without a negation, apply a negation and push down
+            NegationState state = (NegationState) super.visit(node, new NegationState(true));
+            
+            // if the returned state is false the negation was successfully pushed down, it should be removed
+            if (!state.isNegated()) {
+                // drop the not because it was pushed down and get the new node
+                JexlNode nextNode = dropNot(node, new NegationState(true));
+                
+                // the negation at this level was removed so the state is false going forward
+                nextNode.jjtAccept(this, new NegationState(false));
+            }
+            
+            // any down-tree state manipulation is a result of this NOT so do not back propagate that
+            return data;
+        }
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        if (data instanceof NegationState) {
```

Comment:
data is never not an instance of NegationState, right?

Reply:
Did you get a chance to look at this?

Reply:
If using the static `pushdownNegations(query)` method that is true, however its a public interface so this is to be defensive.

Reply:
fair enough.